compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/optimizer/LargeMethodOptimizer.java

```diff
@@ -653,21 +673,40 @@ private BIRFunction createNewBIRFunctionAcrossBB(BIRFunction parentFunc, Name fu
         BIRBasicBlock exitBB = new BIRBasicBlock(newBBNum);
         exitBB.terminator = new BIRTerminator.Return(null);
         lastBB.terminator = new BIRTerminator.GOTO(null, exitBB, lastIns.scope);
-        for (int i = basicBlocks.size() - 1; i >= 0; i--) {
-            BIRTerminator terminator = basicBlocks.get(i).terminator;
-            if (terminator.kind == InstructionKind.GOTO &&
-                    ((BIRTerminator.GOTO) terminator).targetBB.number == lastBBIdNum) {
-                ((BIRTerminator.GOTO) terminator).targetBB = lastBB;
-            } else if (terminator.thenBB != null && terminator.thenBB.number == lastBBIdNum) {
-                terminator.thenBB = lastBB;
-            }
+        for (BIRBasicBlock basicBlock : basicBlocks) {
+            fixTerminatorBBs(lastBBIdNum, lastBB, basicBlock.terminator);
         }
         basicBlocks.add(lastBB);
         basicBlocks.add(exitBB);
         rectifyVarKindsAndTerminators(birFunc, selfVarDcl, exitBB);
         return birFunc;
     }
 
+    private void fixTerminatorBBs(int lastBBIdNum, BIRBasicBlock lastBB, BIRTerminator terminator) {
+        if (terminator.thenBB != null && terminator.thenBB.number == lastBBIdNum) {
+            terminator.thenBB = lastBB;
+        }
+
+        switch (terminator.getKind()) {
+            case GOTO:
+                if (((BIRTerminator.GOTO) terminator).targetBB.number == lastBBIdNum) {
+                    ((BIRTerminator.GOTO) terminator).targetBB = lastBB;
+                }
+                break;
+            case BRANCH:
+                BIRTerminator.Branch branchTerminator = (BIRTerminator.Branch) terminator;
+                if (branchTerminator.trueBB.number == lastBBIdNum) {
+                    branchTerminator.trueBB = lastBB;
+                }
+                if (branchTerminator.falseBB.number == lastBBIdNum) {
+                    branchTerminator.falseBB = lastBB;
+                }
+                break;
+            default:
```

Comment:
Shall we make GOTO as default?

Reply:
Will change if the default case is not hit by the tests.

Reply:
We need to keep the default case to not do anything because all terminators come to this method. Hence keeping this like this not to have ClassCastException.